File: src/main/java/net/ccbluex/liquidbounce/utils/misc/Animation.java
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.utils.misc;

import net.ccbluex.liquidbounce.utils.timer.TimerUtils;

public abstract class Animation  {
    public TimerUtils timerUtil;
    protected int duration;
    protected double endPoint;
    protected Direction direction;
    
    public Animation(final int ms, final double endPoint) {
        this.timerUtil = new TimerUtils();
        this.duration = ms;
        this.endPoint = endPoint;
        this.direction = Direction.FORWARDS;
    }
    
    public Animation(final int ms, final double endPoint, final Direction direction) {
        this.timerUtil = new TimerUtils();
        this.duration = ms;
        this.endPoint = endPoint;
        this.direction = direction;
    }
    
    public boolean finished(final Direction direction) {
        return this.isDone() && this.direction.equals(direction);
    }
    
    public double getLinearOutput() {
        return 1.0 - this.timerUtil.getTime() / (double)this.duration * this.endPoint;
    }
    
    public double getEndPoint() {
        return this.endPoint;
    }
    
    public void setEndPoint(final double endPoint) {
        this.endPoint = endPoint;
    }
    
    public void reset() {
        this.timerUtil.reset();
    }
    
    public boolean isDone() {
        return this.timerUtil.hasTimeElapsed(this.duration);
    }
    
    public void changeDirection() {
        this.setDirection(this.direction.opposite());
    }
    
    public Direction getDirection() {
        return this.direction;
    }
    
    public void setDirection(final Direction direction) {
        if (this.direction != direction) {
            this.direction = direction;
            this.timerUtil.setTime(System.currentTimeMillis() - (this.duration - Math.min(this.duration, this.timerUtil.getTime())));
        }
    }
    
    public void setDuration(final int duration) {
        this.duration = duration;
    }
    
    protected boolean correctOutput() {
        return false;
    }
    
    public double getOutput() {
        if (this.direction == Direction.FORWARDS) {
            if (this.isDone()) {
                return this.endPoint;
            }
            return this.getEquation((double)this.timerUtil.getTime()) * this.endPoint;
        }
        else {
            if (this.isDone()) {
                return 0.0;
            }
            if (this.correctOutput()) {
                final double revTime = (double)Math.min(this.duration, Math.max(0L, this.duration - this.timerUtil.getTime()));
                return this.getEquation(revTime) * this.endPoint;
            }
            return (1.0 - this.getEquation((double)this.timerUtil.getTime())) * this.endPoint;
        }
    }
    
    protected abstract double getEquation(final double p0);
}

File: src/main/java/net/ccbluex/liquidbounce/utils/animations/Animation.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.utils.animations

import net.ccbluex.liquidbounce.utils.timing.MSTimer
import kotlin.math.max
import kotlin.math.min

/**
 * Unified Animation System
 * Replaces all duplicate animation systems across the project
 */
abstract class Animation @JvmOverloads constructor(
    var duration: Int,
    var endPoint: Double,
    direction: Direction = Direction.FORWARDS
) {

    val timerUtil = MSTimer()

    var direction: Direction = direction
        set(newDirection) {
            if (field != newDirection) {
                field = newDirection
                timerUtil.time = System.currentTimeMillis() - (duration.toLong() - min(duration.toLong(), timerUtil.getElapsedTime()))
            }
        }

    val isDone: Boolean
        get() = timerUtil.hasTimeElapsed(duration.toLong())

    val output: Double
        get() = if (direction == Direction.FORWARDS) {
            if (isDone) endPoint
            else getEquation(timerUtil.getElapsedTime().toDouble()) * endPoint
        } else {
            if (isDone) 0.0
            else {
                if (correctOutput()) {
                    val revTime = min(duration.toDouble(), max(0.0, duration.toDouble() - timerUtil.getElapsedTime().toDouble()))
                    getEquation(revTime) * endPoint
                } else {
                    (1 - getEquation(timerUtil.getElapsedTime().toDouble())) * endPoint
                }
            }
        }

    val linearOutput: Double
        get() = 1 - timerUtil.getElapsedTime() / duration.toDouble() * endPoint

    fun setDirection(direction: Direction): Animation {
        this.direction = direction
        return this
    }

    fun finished(direction: Direction): Boolean = isDone && this.direction == direction

    fun reset() = timerUtil.reset()

    fun changeDirection() {
        direction = direction.opposite()
    }

    protected open fun correctOutput(): Boolean = false

    protected abstract fun getEquation(x: Double): Double
}
